pkg/tag/generate_tag_definitions.py

```diff
@@ -1,4042 +1,66 @@
-            ok = False
-            continue
-
-        vr=m.group(3).upper()
-        if vr == "XS":
-            # Its generally safe to treat XS as unsigned.  See
-            # https://github.com/dgobbi/vtk-dicom/issues/38 for
-	    # some discussions.
-            vr = "US"
-        elif vr == "OX":
-	    # TODO(saito) I'm less sure about the OX rule. Where is
-	    # this crap defined in the standard??
-            vr = "OW"
-
-        tag = Tag(group=m.group(1),
-                  elem=m.group(2),
-                  vr=vr,
-                  name=m.group(4),
-                  vm=m.group(5))
-
-
-        if not re.match('^[0-9A-Fa-f]+$', tag.group) or not re.match('^[0-9A-Fa-f]+$', tag.elem):
-            continue
-        tags.append(tag)
-    if not ok:
-        sys.exit(1)
-
-    return tags
+    ('vm', str),
+    ('keyword', str),
+    ('retired', bool)])
+
+def read_tags_from_innolitics(version_hash: str) -> List[Tag]:
+    response = urllib.request.urlopen(f"https://raw.githubusercontent.com/innolitics/dicom-standard/{version_hash}/standard/attributes.json")
+    attrs = json.loads(response.read().decode())
+    allowed_vrs_separator = " or "
+    return [
+        Tag(
+            # The id field should always follow format "ggggeeee", so this should be safe.
+            group=int(resolvable_tag_id[:4], 16),
+            elem=int(resolvable_tag_id[4:], 16),
+            # To understand this ternary expression, see: https://dicom.nema.org/medical/dicom/2024a/output/html/part06.html#note_6_2.
+            vr=(e["valueRepresentation"] if resolvable_tag_id[:4].lower() != 'fffe' else "NA").split(allowed_vrs_separator),
+            name=e["name"],
+            vm=e["valueMultiplicity"],
+            keyword=e["keyword"],
+            retired=e["retired"] == "Y")
+        for e in attrs
+        if (resolvable_tag_id := e["id"].replace("x", "0")) and len(e["keyword"]) > 0
+    ]
 
 def generate(out: IO[str]):
-    tags = list_tags()
-
-    print("package dicomtag", file=out)
-    print("", file=out)
-    print("// Code generated from generate_tag_definitions.py. DO NOT EDIT.", file=out)
-    for t in tags:
-        if t.name.find("RETIRED") >= 0:
-            continue
-        print(f'var {t.name} = Tag{{0x{t.group}, 0x{t.elem}}}', file=out)
-
-    print("var tagDict map[Tag]TagInfo", file=out)
-    print("", file=out)
-    print("func init() {", file=out)
-    print("	maybeInitTagDict()", file=out)
-    print("}", file=out)
-    print("func maybeInitTagDict() {", file=out)
-    print("	if len(tagDict) > 0 {", file=out)
-    print("		return", file=out)
-    print("	}", file=out)
-    print("	tagDict = make(map[Tag]TagInfo)", file=out)
-    for t in tags:
-        print(f'	tagDict[Tag{{0x{t.group}, 0x{t.elem}}}] = TagInfo{{Tag{{0x{t.group}, 0x{t.elem}}}, "{t.vr}", "{t.name}", "{t.vm}"}}', file=out)
-    print("}", file=out)
+    tags = read_tags_from_innolitics(INNOLITICS_VERSION_HASH)
+    new_line = chr(10)
```

Comment:
out of curiosity is there a reason to do this instead of `'\n'`? Is this more portable somehow? If not, I'd prefer \n for readability.

Reply:
For some reason you can't have backslashes in an f-string. So it failed when I tried `'\n'. See https://stackoverflow.com/questions/44780357/how-can-i-use-newline-n-in-an-f-string-to-format-output

Apparently this issue is fixed in Python 3.12 it works.